src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,115 @@
+		PhysicalPlan bestPlan = initialPlan;
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel(); //Created as a value and stored rather than just using the getCostModel, because the same model is used later too.
+
+		Double[] initialCost = CostEstimationUtils.getEstimates(costModel, initialPlan);
+		
+		// generation = number of times the outer loop has run. Has to be declared here since it will increment each outer loop.
+		int generation = 0;
+		
+		// bestCost = best possible cost out of everything that the algorithm has found. Has to be declared here so that we have something to compare to in all the iterations of the outer loop without losing track of it between loops.
+		Double bestCost = initialCost[0];
+		
+		while(!condition.readyToStop(generation)) { // Currently only handles generation number as a stopping condition!
+			
+			// The randomized plan generator is to be used here. As a temporary measure, the initial plan is used.
+			PhysicalPlan currentPlan = initialPlan; // This variable will hold the plan which is currently being worked on.
+			Double currentCost = initialCost[0]; // The cost of the current plan. For it to be a local minimum, none of its neighbours can have a lower cost
+			boolean improvementFound = false;
+			
+			do {
+				final List<PhysicalPlan> neighbours = getNeighbours(currentPlan);
+				final Double[] neighbourCosts = CostEstimationUtils.getEstimates(costModel, neighbours);
+				final List<Pair<PhysicalPlan,Double>> betterPlans = new ArrayList<>();
+				improvementFound = false;
+				
+				// Using a for-loop in order to have the index for which neighbouring plan to pick.
+				for (int x = 0; x < neighbourCosts.length; x++) {
+					if(neighbourCosts[x] < currentCost) {
+						improvementFound = true;
+						final Pair<PhysicalPlan,Double> betterPlan = new Pair<PhysicalPlan, Double>(neighbours.get(x), neighbourCosts[x]);
+						betterPlans.add(betterPlan); // I haven't figured out how to do this properly for Java yet, but it will be solved by the PhysicalPlanWithCost anyway.
+					}
+				}
+				
+				if(improvementFound) { // if we have found at least one possible improvement, we want to make one of them into our new current plan.
+					final Pair<PhysicalPlan,Double> newPlan = betterPlans.get(rng.nextInt(betterPlans.size())); // Get a random object.
+					currentPlan = newPlan.object1;
+					currentCost = newPlan.object2;
+				}
+			}
+			while(improvementFound);
+			
+			if(currentCost < bestCost) {
+				bestCost = currentCost;
+				bestPlan = currentPlan;
+			}
+			
+			generation++;
+		}
+		
+		return bestPlan;
+	}
+	
+	protected List<PhysicalPlan> getNeighbours(final PhysicalPlan initialPlan) {
+		List<PhysicalPlan> resultList = new ArrayList<PhysicalPlan>(); // Create an empty list. Just List<PhysicalPlan> didn't work so I had to look this up.
+		
+		RuleInstances rInst = new RuleInstances();

```

Comment:
Make `rInst` a member variable (like `rng`).

Reply:
In fact, make the class `RuleApplicationsOfPlans` the member variable and then use that one in the following.

Reply:
To be a bit more precise here, I meant that the type of the member variable should be `RuleApplicationsOfPlans`. Something like the following:
```
   protected final RuleApplicationsOfPlans rules = new RuleApplicationsOfPlans( new RuleInstances() );
```